scripts/core/editor3/helpers/highlights.js

```diff
@@ -0,0 +1,271 @@
+
+export function getHighlightsState(editorState) {
+    return getCustomDataFromEditor(editorState, editor3DataKeys.MULTIPLE_HIGHLIGHTS)
+        || getInitialHighlightsState();
+}
+
+export function setHighlightsState(editorState, hightlightsState) {
+    return setCustomDataForEditor(editorState, editor3DataKeys.MULTIPLE_HIGHLIGHTS, hightlightsState);
+}
+
+function getHighlightTypeFromStyle(style) {
+    const delimiterIndex = style.lastIndexOf('-');
+
+    if (delimiterIndex === -1) {
+        null;
+    }
+
+    const type = style.slice(0, delimiterIndex);
+
+    if (Object.keys(availableHighlights).includes(type)) {
+        return type;
+    }
+
+    return null;
+}
+
+/**
+ * @ngdoc method
+ * @name getHighlightStyleAtOffset
+ * @param {Object} editorState
+ * @param {List} types
+ * @param {Object} selection
+ * @param {Integer} offset
+ * @description the entity key from the new possition specified by offset.
+ */
+export function getHighlightStyleAtOffset(editorState, types, selection, offset) {
+    const content = editorState.getCurrentContent();
+    const {block, newOffset} = getBlockAndOffset(content, selection, offset);
+
+    if (block == null) {
+        return null;
+    }
+
+    const inlineStyles = block.getInlineStyleAt(newOffset);
+    let highlightStyle = null;
+
+    inlineStyles.forEach((style) => {
+        const type = getHighlightTypeFromStyle(style);
+
+        if (type != null && types.indexOf(type) !== -1) {
+            highlightStyle = style;
+        }
+    });
+
+    return highlightStyle;
+}
+
+export function getHighlightStyle(editorState, types) {
+    const selection = editorState.getSelection();
+
+    return getHighlightStyleAtOffset(editorState, types, selection, 0);
+}
+
+/**
+ * @ngdoc method
+ * @name getHighlightData
+ * @param {Object} editorstate
+ * @return {String} style
+ * @description returns the data associated to the style.
+ */
+export function getHighlightData(editorState, style) {
+    const highlightsState = getHighlightsState(editorState);
+
+    return highlightsState.highlightsData[style];
+}
+
+/**
+ * @ngdoc method
+ * @name getHighlightDataAtOffset
+ * @param {Object} content
+ * @param {Object} selection
+ * @param {Integer} offset
+ * @return {String} entity key
+ * @description the entity key from the new possition specified by offset.
+ */
+export function getHighlightDataAtOffset(editorState, types, selection, offset) {
+    const style = getHighlightStyleAtOffset(editorState, types, selection, offset);
+
+    if (style == null) {
+        return null;
+    }
+
+    const highlightsState = getHighlightsState(editorState);
+
+    return highlightsState.highlightsData[style];
+}
+
+/**
+ * @ngdoc method
+ * @name addHighlightData
+ * @param {Object} editorState
+ * @param {String} type
+ * @param {Object} data
+ * @return {Object} new editor state
+ * @description add a new highlight at the cussrnt selection.
+ */
+export function addHighlightData(editorState, type, data) {
```

Comment:
Removing a visual highlight in the editor and the associated data should be done in a single transaction, so we don't end up with a state when a visual highlight is removed and data isn't or vice versa.

Also, it duplicates the code at `MultipleHighlights.jsx > _addHighlight`

Reply:
The transaction re: issue is not present here.
Yes, the highlightsManager is not yet stateless. More refactoring is needed on MultipleHighlights in order to achieve this. I will be done in next commit for this PR.